File: iterator/java/Iterator.java
// Book
class Book {
    private String title;
    private String author;

    public Book(String title, String author) {
        this.title = title;
        this.author = author;
    }

    public String getTitle() {
        return title;
    }

    public String getAuthor() {
        return author;
    }
}

// Iterator interface
interface Iterator {
    boolean hasNext();
    Book next();
}

// Concrete Iterator
class BookIterator implements Iterator {
    private Book[] books;
    private int index = 0;

    public BookIterator(Book[] books) {
        this.books = books;
    }

    @Override
    public boolean hasNext() {
        return index < books.length;
    }

    @Override
    public Book next() {
        if (hasNext()) {
            return books[index++];
        }
        return null;
    }
}

// Collection
class BookCollection {
    private Book[] books;
    private int count = 0;

    public BookCollection(int size) {
        books = new Book[size];
    }

    public void addBook(Book book) {
        if (count < books.length) {
            books[count++] = book;
        }
    }

    public Iterator createIterator() {
        return new BookIterator(books);
    }
}

// Main class demonstrating Iterator pattern
public class Iterator {
    public static void main(String[] args) {
        BookCollection collection = new BookCollection(3);
        collection.addBook(new Book("Design Patterns", "Gang of Four"));
        collection.addBook(new Book("Clean Code", "Robert Martin"));
        collection.addBook(new Book("The Pragmatic Programmer", "Hunt & Thomas"));

        iterator.Iterator iterator = collection.createIterator();
        System.out.println("Books in collection:");
        while (iterator.hasNext()) {
            Book book = iterator.next();
            if (book != null) {
                System.out.println("- " + book.getTitle() + " by " + book.getAuthor());
            }
        }
    }
}

File: iterator/typescript/Iterator.ts
// Book
class Book {
  title: string;
  author: string;

  constructor(title: string, author: string) {
    this.title = title;
    this.author = author;
  }
}

// Iterator interface
interface Iterator<T> {
  hasNext(): boolean;
  next(): T | null;
}

// Concrete Iterator
class BookIterator implements Iterator<Book> {
  private books: Book[];
  private index: number = 0;

  constructor(books: Book[]) {
    this.books = books;
  }

  hasNext(): boolean {
    return this.index < this.books.length;
  }

  next(): Book | null {
    if (this.hasNext()) {
      return this.books[this.index++];
    }
    return null;
  }
}

// Collection
class BookCollection {
  private books: Book[] = [];

  addBook(book: Book): void {
    this.books.push(book);
  }

  createIterator(): Iterator<Book> {
    return new BookIterator(this.books);
  }
}

// Usage
const collection = new BookCollection();
collection.addBook(new Book('Design Patterns', 'Gang of Four'));
collection.addBook(new Book('Clean Code', 'Robert Martin'));
collection.addBook(new Book('The Pragmatic Programmer', 'Hunt & Thomas'));

const iterator = collection.createIterator();
console.log('Books in collection:');
while (iterator.hasNext()) {
  const book = iterator.next();
  if (book) {
    console.log(`- ${book.title} by ${book.author}`);
  }
}
